Fix decoding bug in C implementation

gonzus opened the pull request:
Bug was due to two things (so, maybe two bugs):

1. Not considering the full size of a 16 characters code.
2. Not correcting the top position when the code includes a separator,
   which made it process fewer grid square characters.

A new test was added to make sure this is (and remains) fixed.

drinckes commented:
Hi @gonzus  - could you change the test code so that it reads the tests from the test_data files? That way, all the implementations get the same tests and we know they are consistent.

gonzus commented:
Hi @drinckes -- it does already. It tries to read data from `../test_data` and `./test_data`, to allow running the tests inside the `c` directory, and also from the root directory as `make -Cc clean test`.

gonzus commented:
Oh, if you mean the particular test I added for this PR... I didn't see clearly how to add the test to the data files, so the values are hardcoded in the C library test.

drinckes commented:
Oh sorry about that, didn't look closely enough.

The test you added checks that encoding to 16 digits produces "8FVC2222+22GCCCC", yes?

For code length, the "+" character isn't counted - the length is the number of _digits_, so "8FVC2222+22GCCCC" is considered to have length **15**.

gonzus commented:
Yes to your first question.
I just pushed a commit that explicitly tests the code length being 15 for that case.

drinckes commented:
My point is you request the code with length 16. What you are generating is too short because of the changes in `olc.c`.

gonzus commented:
Ah, now I see. Yes, you are right. I changed the parameter to `15` (test still passes, of course) and added a couple of comments regarding what `OLC_Encode` and `OLC_EncodeDefault` return.

drinckes commented:
I've just run `make test` and get the following errors, could you check it builds?

(I'm using gcc version 7.3.0 (Debian 7.3.0-5).)

```
c$ make test
cc -c -std=c99 -g -Wall -Wno-comment  -o olc_test.o olc_test.c
In file included from olc_test.c:6:0:
olc_private.h:13:42: error: initializer element is not constant
 static const size_t kGridRows          = kEncodingBase / kGridCols;
                                          ^~~~~~~~~~~~~
olc_private.h:21:42: error: initializer element is not constant
 static const double kLatMaxDegreesT2   = 2 * kLatMaxDegrees;
                                          ^
olc_private.h:26:42: error: initializer element is not constant
 static const double kLonMaxDegreesT2   = 2 * kLonMaxDegrees;
                                          ^
olc_test.c: In function ‘to_boolean’:
olc_test.c:135:9: warning: implicit declaration of function ‘strcasecmp’; did you mean ‘strncmp’? [-Wimplicit-function-declaration]
     if (strcasecmp(s, "false") == 0 ||
         ^~~~~~~~~~
         strncmp
In file included from olc_test.c:6:0:
At top level:
olc_private.h:31:15: warning: ‘kInitialResolutionDegrees’ defined but not used [-Wunused-variable]
 static double kInitialResolutionDegrees = 0.0;
               ^~~~~~~~~~~~~~~~~~~~~~~~~
olc_private.h:30:15: warning: ‘kGridSizeDegrees’ defined but not used [-Wunused-variable]
 static double kGridSizeDegrees          = 0.0;
               ^~~~~~~~~~~~~~~~
olc_private.h:29:15: warning: ‘kInitialExponent’ defined but not used [-Wunused-variable]
 static size_t kInitialExponent          = 0;
               ^~~~~~~~~~~~~~~~
Makefile:23: recipe for target 'olc_test.o' failed
make: *** [olc_test.o] Error 1
```

gonzus commented:
I run `make test` before every push, so yes, I can confirm it builds on my end.

Looking at the errors, these are clearly down to the OS / compiler / library. I could try to make this more portable, but the best shot would be for you to send a patch that makes things work on your end.

Comments regarding the specific errors:
* `error: initializer element is not constant`: I would guess the C compiler you are using cannot deduce that the initializer is, in fact, constant. One possibility here is to use precompiler macros for some / all of the constants.
* `warning: implicit declaration of function ‘strcasecmp’`: typical portability issue, this function has a different name on different platforms (`strcmpi()`, `stricmp()`, something else?)

Just in case it is useful, here is what I am using:
```bash
$ cc --version
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

$ uname -a
Darwin C02TM1G3FVH4 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
```

drinckes commented:
Re `initializer is not constant` errors. It appears that the recommended way to define constants in C (as opposed to C++) is to use `#define` (or `enum`).

The `strcasecmp` warning depends on the `-std=c99` flag in the `Makefile` - that flag forces the C99 standard which doesn't include `strcasecmp`. When you run the tests is it using the `Makefile` or is it using settings in your XCode setup? Can you see what the compiler command is when it runs the test or you run `make lib`?

gonzus commented:
I just pushed a commit that made it compile on one of my linux boxes. Can you try again?

I was trying to make the code look similar to C++, hence the compiler constants instead of preprocessor macros.

I added `#include <strings.h>` to fix the `strcasecmp()` thing.

gonzus commented:
Here is a full log of how I run the tests:
```
$ make -Cc clean all test
make: Entering directory `/home/gonzus/src/open-location-code/c'
rm -f crash-* slow-unit-* *.dSYM
rm -f olc.o libolc.a
rm -f example.o olc_test.o
rm -f example olc_test
rm -f olc_test
cc -c -std=c99 -g -Wall -Wno-comment  -o olc.o olc.c
ar rvs libolc.a olc.o
ar: creating libolc.a
a - olc.o
cc -c -std=c99 -g -Wall -Wno-comment  -o olc_test.o olc_test.c
cc -std=c99 -g -o olc_test olc_test.o -lm -L. -lolc
./olc_test 2>&1 | egrep -v '^OK '
            shortCodeTests.csv =>  17 records,  17 OK,   0 BAD
             encodingTests.csv =>  18 records,  18 OK,   0 BAD
             validityTests.csv =>  17 records,  17 OK,   0 BAD
make: Leaving directory `/home/gonzus/src/open-location-code/c'
```

drinckes commented:
Tests pass now so merging! Woooooooooooooooo!
